### PR TITLE
22127: check that products is set

### PR DESCRIPTION
**Pull request by davidverholen**
Magento 2.3.0: getSize call on configurable collection leads to exception, if no product filters are applied

### Description (*)
The newrelic cron tries to get the amount of configurable products (or at least the given children) calling the getCount function of the ConfigurableProductManagementInterface.
In this process a configurable collection object is created (magento/module-configurable-product/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php).
There the getSize function (in AbstractDB parent class) is called and as a consequence the _renderFilters function in the magento/module-configurable-product/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php class.

### Fixed Issues (if relevant)

1. https://github.com/magento/magento2/issues/22127

### Manual testing scenarios (*)

1. Run magento_newrelicreporting_cron cron job

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by davidverholen** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @davidverholen. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @davidverholen, here is your [new Magento instance](https://pr-22186.instances.magento-community.engineering).
Admin access: [https://pr-22186.instances.magento-community.engineering/admin](https://pr-22186.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: What if `$parentIds` is an empty array?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: not sure what select would do, but yes, it would totally make sense to move it into the if because it is empty anyway when the if is skipped

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]:
```suggestion
        if (!$this->products) {
```

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (orlangur) [inline app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: No reason to use `array_map` here.

**Review by orlangur — CHANGES_REQUESTED** [review]
After changes are applied and all builds are green, please squash them into a single commit so that we have perfectly clean history 😉

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: array_map results in less verbose code when you want to create a new array of same size. I see no reason not to use it tbh

Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: plus you gain type safety for array elements

Comment (orlangur) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: Code became much less simpler, please revert.

Type safety is not a real issue and does not justify decrease of readability.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: @orlangur  I disagree with that. It seems to be a very subjective opinion of you

1. array_map is much clearer in its intention: "map values of an array to a new array" .
2. type safety is always better in regards of potential future bugs
3. the type clearly states with which object your dealing (better readability)
4. with this awful value=[]; foreach... approach you have a code block that has to stay together, but is not bound together by syntax

Comment (orlangur) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]:
```
        $parentIds = [];	           
        foreach ($this->products as $product) {	
            $parentIds[] = $product->getData($metadata->getLinkField());	
        }
```

is MUCH easier to comprehend and there is nothing subjective with that. 

Please just fix a bug and revert anything unrelated. For using `array_map` instead of such a simple loop you can create a proposal in https://github.com/magento/architecture.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by orlangur** [comment]
For type safety simply add type hint here: https://github.com/magento/magento2/blob/90d4bccced95603e1c380473ae0b4af904d8b03c/app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php#L66 It must not be enforced in some random places, check in setter in enough.

File: app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php
Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: it's subjective as long as you cannot explain, why it is easier

Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: plus it's not an architecture decision, it's just about clean, readable code

Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: for further reading: https://szymonkrajewski.pl/why-you-should-use-array-functions-instead-of-basic-loops/

Comment (orlangur) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: @davidverholen `foreach` I mentioned is a clean, readable code. Please add at least SOME argument why such `foreach` is bad. Your code is not clean and not readable at all.

There is nothing new for me in that article, I know all about such functions and there are good use cases for them but this is not the one.

If you want to perform such kind of useless refactoring, you need to get an architectural approval first and add recommendation somewhere in https://devdocs.magento.com/guides/v2.3/coding-standards/technical-guidelines.html which would require any suitable loop to be written in form of `array_map`.

This PR is a bugfix and your "refactoring" is not related to achieving a goal. That's why I'm asking you to revert unrelated changes and deliver suggested improvements in separate PR once they get approved.

Comment (davidverholen) [reply app/code/Magento/ConfigurableProduct/Model/ResourceModel/Product/Type/Configurable/Product/Collection.php]: I provided several arguments why array_map is superior in that context, while you provided exactly 0

**Comment by davidverholen** [comment]
@orlangur it's not a random place, it's where the object is actually used and where from a developer experience and stability perspective it's a total benefit to have type safety.

You still did not bring one real argument against using array_map here (besides you obviously not liking it).

So why do we have this discussion?

**Comment by josevectorche** [comment]
Sure thanks for the event.

**Comment by arnoudhgz** [comment]
@orlangur @davidverholen is it not more convenient and better to understand if you just set the `$this->products` variable default to an empty array since it is already a `private` variable?

Also on the `setProductFilter` method it expects already an array.

**Comment by davidverholen** [comment]
@arnoudhgz I like the Idea, because it's very simple and maybe we can then finally close this pr. Thx for the suggestion, I will change it

**Review by orlangur — APPROVED** [review]
@davidverholen thanks for getting rid of unrelated code refactoring 👍 

@arnoudhgz nice catch, I overlooked such simplification 🙇

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-5099](https://jira.corp.magento.com/browse/ENGCOM-5099) has been created to process this Pull Request

**Comment by soleksii** [comment]
:heavy_check_mark: QA Passed

**Before:**

before

**After:**

after

**Comment by raghuiamsingh** [comment]
How can I apply this change to a Magento Installation on v2.3.2?

**Pull request merged**
